types/lib/model.d.ts

```diff
@@ -1825,25 +1830,43 @@ export abstract class Model<TModelAttributes extends {} = any, TCreationAttribut
   public static findByPk<M extends Model>(
     this: ModelStatic<M>,
     identifier: Identifier,
-    options: Omit<NonNullFindOptions<M['_attributes']>, 'where'>
+    options: Omit<NonNullFindOptions<M['_attributes']>, 'where'> & { raw?: false }
   ): Promise<M>;
+  public static findByPk<M extends Model, R extends any = M['_attributes']>(
+    this: ModelStatic<M>,
+    identifier: Identifier,
+    options: Omit<NonNullFindOptions<M['_attributes']>, 'where'> & { raw: true }
+  ): Promise<R>;
   public static findByPk<M extends Model>(
     this: ModelStatic<M>,
     identifier?: Identifier,
-    options?: Omit<FindOptions<M['_attributes']>, 'where'>
+    options?: Omit<FindOptions<M['_attributes']>, 'where'> & { raw?: false }
   ): Promise<M | null>;
+  public static findByPk<M extends Model, R extends any = M['_attributes']>(
+    this: ModelStatic<M>,
+    identifier?: Identifier,
+    options?: Omit<FindOptions<M['_attributes']>, 'where'> & { raw: true }
+  ): Promise<R | null>;
```

Comment:
`plain: true` isn't being handled and I think you can do this in a neater way.

eg:
```ts
type ModelAttributes<TModel extends Model> = TModel['_attributes'];

  public static findByPk<
    M extends Model,
    TOptions extends Omit<
      FindOptions<ModelAttributes<M>>, 'where'
  >>(
    this: ModelStatic<M>,
    identifier?: Identifier,
    options?: TOptions
  ): Promise<
    null
    | TOptions extends { raw: true } | { plain: true } 
      ? ModelAttributes<M>
      : M
    > {
```

TS might not be smart enough to handle this though, not sure.

Reply:
This conversation should occur on the PR this is based on https://github.com/sequelize/sequelize/pull/12921/files

But for the sake of response...
To my knowledge Sequelize doesn't advertise the `plain` option for Model lookups... so I wouldn't expect special type handling for that use case.

The only documentation I can find is for raw SQL queries+custom getters, both of which have custom typing for this use case.

Also that suggested change removes the custom typing for `raw` which is the point of that PR

Reply:
Nope, it still has the raw part, though I'm not sure if it'll work: depends on how smart ts is when it infers the typing from TOptions. It'll either be have `raw?: boolean` or the actual type of the value (might be able to use `as const` to get ts to infer it better, but I haven't looked into that.

The `TOptions extends { raw: true } | { plain: true } ? ModelAttributes<M> : M` will give the model attributes if `TOptions` has `raw: true` or `{ plain: true }` and fallback on M. 

Mostly just a matter of preference.